src/Features/Lsif/Generator/ResultSetTracking/SymbolHoldingResultSetTracker.cs

```diff
@@ -0,0 +1,202 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using Microsoft.CodeAnalysis.LanguageServerIndexFormat.Generator.Graph;
+using Microsoft.CodeAnalysis.LanguageServerIndexFormat.Generator.Writing;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.LanguageServerIndexFormat.Generator.ResultSetTracking
+{
+    internal sealed class SymbolHoldingResultSetTracker : IResultSetTracker
+    {
+        private readonly Dictionary<ISymbol, TrackedResultSet> _symbolToResultSetId = new Dictionary<ISymbol, TrackedResultSet>();
+        private readonly ReaderWriterLockSlim _readerWriterLock = new ReaderWriterLockSlim(LockRecursionPolicy.NoRecursion);
+        private readonly ILsifJsonWriter _lsifJsonWriter;
+
+        /// <summary>
+        /// The compilation which we are analyzing. When we make ResultSets, we attach monikers to them, and those
+        /// monikers express an import/export concept for symbols being consumed from another project. We must distinguish
+        /// from source symbols that come from the project being analyzed versus symbols from referenced compilations, so
+        /// we can't just use <see cref="ISymbol.Locations" /> to make the determination.
+        /// </summary>
+        private readonly Compilation _sourceCompilation;
+
+        public SymbolHoldingResultSetTracker(ILsifJsonWriter lsifJsonWriter, Compilation sourceCompilation)
+        {
+            _lsifJsonWriter = lsifJsonWriter;
+            _sourceCompilation = sourceCompilation;
+        }
+
+        private TrackedResultSet GetTrackedResultSet(ISymbol symbol)
+        {
+            TrackedResultSet trackedResultSet;
+
+            // First acquire a simple read lock to see if we already have a result set; we do this with
+            // just a read lock to ensure we aren't contending a lot if the symbol already exists which
+            // is the most common case.
+            using (_readerWriterLock.DisposableRead())
+            {
+                if (_symbolToResultSetId.TryGetValue(symbol, out trackedResultSet))
+                {
+                    return trackedResultSet;
+                }
+            }
+
+            using (var upgradeableReadLock = _readerWriterLock.DisposableUpgradeableRead())
```

Comment:
An upgradeable read prevents other upgradeable reads from entering the lock at the same time, so it has the same exclusion behavior that way, but it still allows other reads to either enter or at least leave. I forget the precise details (the docs are very precise) but the goal was to avoid having to acquire the outright write lock until we're absolutely positive we need it.

I didn't benchmark without it but given the ReaderWriterLockSlim supports this it seemed easy enough.